### ISSUE TITLE
Track field usage

### ISSUE THREAD

**Issue by danielvdao** [rmosolgo/graphql-ruby#4055]
**Is your feature request related to a problem? Please describe.**
The TL;DR is that managing deprecated fields / code in our graphql-ruby backend becomes hard the larger we scale. Eventually clients stop using deprecated fields, but figuring out **when** is the problem. Because of this, developers tend to get hesitant to remove legacy code as it can accidentally trigger old legacy issues.

I'm curious to hear other's solutions to this, or if there's something that we can come up with together 😓  

Reading [this blog](https://netflixtechblog.com/how-netflix-scales-its-api-with-graphql-federation-part-2-bbe71aaec44a) by Netflix - I saw the following:
> We have a deprecation workflow in place for evolving the schema. We’ve leveraged GraphQL’s [deprecation](https://spec.graphql.org/June2018/#sec--deprecated) feature and also track usage stats for every field in the schema. Once the stats show that a deprecated field is no longer used, we can make a backward incompatible change to remove the field from the schema.

Our clients send headers specifying app-version today already and we also track operation name, but finer granularity like field name is not as easy to see. One way I can somewhat see is looking at the `query` instance in `data[:multiplex].queries` and trying to reverse-engineer it through there, but maybe there is a better way.

**Describe the solution you'd like**

The end result of this is that clients can easily increment a metric to track a field usage. Long-term we'd like to fully remove code from our servers, as field deprecation can become bloated over-time on the backend, but having observability into usage is key.

**Describe alternatives you've considered**

I've tried looking around online - but it seems like nobody has a current solution to this. One thing we also do is forced application upgrades based a specific version, but still, this would be much better.

If we could track field usage and couple that data with the number of clients on version, it'd be easier to axe legacy code.

I've tried looking at the `GraphQL::Tracing::DataDogTracing` class to get a better understanding of it -- but was wondering if you happened to know as well.

**Comment by swalkinshaw** [comment]
Check out https://github.com/Shopify/graphql-metrics since it gathers extensive field usage metrics.

**Comment by danielvdao** [comment]
@swalkinshaw thanks for the link. I'll take a look and see if I can whip up a toy example sometime this week and get back to you 🙏🏽

**Comment by rmosolgo** [comment]
This analyzer might provide some inspiration, too: https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/analysis/ast/field_usage.rb Even if it's not quite right out-of-the-box, you could certainly implement your own analyzer for this (docs on AST analysis: https://graphql-ruby.org/queries/ast_analysis.html).


Let me know how it goes if you give it a try -- maybe there's something we could add that would make it more useful to you!

**Comment by danielvdao** [comment]
@swalkinshaw I'd be curious to understand what're some of the way Shopify tackles removing fields if possible? We're starting to track client versions and field usage now (thanks to the gem!).
